ui/details-panel/AutoclassifyTab.jsx

```diff
@@ -0,0 +1,535 @@
+
+        selectedLineIds.add(toAdd.id);
+      } else if (direction === 'previous') {
+        // try to select the previous line.
+        const prevIdx = errorLines.indexOf(firstSelected) - 1;
+        const toAdd = prevIdx >= 0 ? errorLines[prevIdx] : firstSelected;
+
+        selectedLineIds.add(toAdd.id);
+      } else if (direction === 'all_next') {
+        const toAdd = errorLines.slice(errorLines.indexOf(lastSelected));
+        toAdd.forEach(line => selectedLineIds.add(line.id));
+      }
+    } else {
+      const firstUnverified = errorLines.find(line => !line.verified);
+      if (firstUnverified) {
+        selectedLineIds.add(firstUnverified.id);
+      }
+    }
+    this.setState({ selectedLineIds });
+
+    // Scroll the first selected index into view
+    const newFirstSelectedIdx = errorLines.findIndex(line => selectedLineIds.has(line.id));
+    $(".autoclassify-error-lines .error-line")[newFirstSelectedIdx]
+        .scrollIntoView({ behavior: "smooth", block: "start" });
+  }
+
+  getPendingLines() {
+    const { errorLines } = this.state;
+
+    return errorLines.filter(line => !line.verified);
+  }
+
+  getSelectedLines() {
+    const { selectedLineIds, inputByLine } = this.state;
+
+    return Array.from(selectedLineIds).reduce((lines, id) => {
+      const settings = inputByLine.get(id);
+      return settings ? [...lines, settings] : lines;
+    }, []);
+  }
+
+  getLoadStatusText() {
+    switch (this.state.loadStatus) {
+      case 'job_pending': return 'Job not complete, please wait';
+      case 'pending': return 'Logs not fully parsed, please wait';
+      case 'failed': return 'Log parsing failed';
+      case 'no_logs': return 'No errors logged';
+      case 'error': return 'Error showing autoclassification data';
+      case 'loading': return null;
+      case 'ready': return (!this.state.errorLines || this.state.errorLines.length === 0) ? 'No error lines reported' : null;
+      default: return `Unexpected status: ${this.state.loadStatus}`;
+    }
+  }
+
+  setEditable(lineIds, editable) {
+    const { editableLineIds } = this.state;
+    const f = editable ? lineId => editableLineIds.add(lineId) :
+      lineId => editableLineIds.delete(lineId);
+
+    lineIds.forEach(f);
+    this.setState({ editableLineIds });
+  }
+
+  setErrorLineInput(id, input) {
+    const { inputByLine } = this.state;
+
+    inputByLine.set(id, input);
+    this.setState({ inputByLine });
+  }
+
+  /**
+   * Emit an event indicating that the job has been fully classified
+   */
+  signalFullyClassified() {
+    const { job } = this.props;
+
+    // Emit this event to get the main UI to update
+    this.$rootScope.$emit(thEvents.autoclassifyVerified, { jobs: { [job.id]: job } });
+  }
+
+  /**
+   * Get TextLogerror data from the API
+   */
+  async fetchErrorData() {
+    const { job } = this.props;
+
+    if (job.id) {
+      const matcherResp = await fetch(getApiUrl("/matcher/"));
```

Comment:
Did we cache this before? Also can you set this up so that the downloads are started in parallel (dunno fi there's a fancy new way to do it, buy presumably something like
```
const [matcherResp, errorLineResp] = await Promise.all([fetch(getApiUrl("/matcher/")) , fetch(getProjectJobUrl('/text_log_errors/', job.id))]);
```

(but I only wrote that to have a go with the syntax, so it might be all wrong).

Reply:
Actually, your syntax worked quite brilliantly!  :)  But as you alluded to, the matchers should be cached.  So I ended up splitting it into a separate section to load it only once.  With that, I ended up not using the ``Promise.all``.